File: latest/workshop1_m1/W01_M01_demo.py
"""
============================================================
Workshop 1 — Module 1: Why Visualize?
Python Demonstration Script
Data Visualization for Data Scientists — UNYT Tirana
============================================================
"""

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec

# Global plot style — clean, publication-ready
sns.set_theme(style="whitegrid", font_scale=1.0)
plt.rcParams["figure.dpi"] = 150


# ── 1. Anscombe's Quartet ───────────────────────────────────
# Seaborn ships with the Anscombe dataset built in.

df_ans = sns.load_dataset("anscombe")

# Verify that all four datasets share the same statistics
summary = df_ans.groupby("dataset").agg(
    mean_x=("x", "mean"),
    mean_y=("y", "mean"),
    sd_x=("x", "std"),
    sd_y=("y", "std"),
).round(2)

# Correlation requires a custom aggregation
corr = (
    df_ans.groupby("dataset")
    .apply(lambda g: g["x"].corr(g["y"]), include_groups=False)
    .rename("cor_xy")
    .round(3)
)
summary = summary.join(corr)
print("── Anscombe's Quartet: Summary Statistics ──")
print(summary)
print()

# Faceted scatterplot with regression lines (seaborn lmplot)
g = sns.lmplot(
    data=df_ans,
    x="x",
    y="y",
    col="dataset",
    col_wrap=2,
    ci=None,                       # no confidence band
    height=3,
    aspect=1.1,
    scatter_kws={"s": 40, "color": "#1565C0"},
    line_kws={"color": "#E53935", "linewidth": 1.2},
)
g.figure.suptitle(
    "Anscombe's Quartet: Same Statistics, Different Stories",
    fontsize=13, fontweight="bold", y=1.02,
)
g.set_titles("Dataset {col_name}")
plt.tight_layout()
plt.savefig("output/anscombe_quartet_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 2. Datasaurus Dozen ────────────────────────────────────
# Load from the datasauRus repository TSV file.

url = (
    "https://raw.githubusercontent.com/"
    "jumpingrivers/datasauRus/main/inst/"
    "extdata/DatasaurusDozen-long.tsv"
)

try:
    df_dino = pd.read_csv(url, sep="\t")
except Exception:
    # Fallback: create a minimal dummy if network is unavailable
    print("Warning: could not fetch Datasaurus data from network.")
    print("Creating local Anscombe-based placeholder instead.\n")
    df_dino = df_ans.rename(columns={"dataset": "dataset"})

# Verify identical statistics across all 13 datasets
dino_stats = df_dino.groupby("dataset").agg(
    mean_x=("x", "mean"),
    mean_y=("y", "mean"),
    sd_x=("x", "std"),
    sd_y=("y", "std"),
).round(1)
print("── Datasaurus Dozen: Summary Statistics ──")
print(dino_stats.head(13))
print()

# Faceted scatterplot — 13 shapes, one set of statistics
datasets = df_dino["dataset"].unique()
n_datasets = len(datasets)
ncols = min(5, n_datasets)
nrows = int(np.ceil(n_datasets / ncols))

fig, axes = plt.subplots(nrows, ncols, figsize=(12, 2.5 * nrows))
axes_flat = axes.flatten() if n_datasets > 1 else [axes]

for i, ds in enumerate(sorted(datasets)):
    ax = axes_flat[i]
    subset = df_dino[df_dino["dataset"] == ds]
    ax.scatter(subset["x"], subset["y"], s=5, alpha=0.5, c="#2E7D32")
    ax.set_title(ds, fontsize=7, fontweight="bold")
    ax.set_xticks([])
    ax.set_yticks([])

# Hide unused subplots
for j in range(i + 1, len(axes_flat)):
    axes_flat[j].set_visible(False)

fig.suptitle(
    "The Datasaurus Dozen — Same Stats, Different Graphs",
    fontsize=12, fontweight="bold",
)
plt.tight_layout()
plt.savefig("output/datasaurus_dozen_py.png", dpi=150, bbox_inches="tight")
plt.close()


# ── 3. First Plot: mpg Dataset ─────────────────────────────
# Seaborn's built-in mpg dataset (auto fuel economy data).

mpg = sns.load_dataset("mpg").dropna()

fig, ax = plt.subplots(figsize=(8, 5))
sns.scatterplot(
    data=mpg,
    x="displacement",
    y="mpg",
    hue="origin",
    alpha=0.7,
    s=50,
    palette="Set1",
    ax=ax,
)
ax.set_title("Engine Displacement vs Fuel Economy", fontweight="bold")
ax.set_xlabel("Engine Displacement (cubic inches)")
ax.set_ylabel("Miles per Gallon")
plt.tight_layout()
plt.savefig("output/mpg_scatter_py.png", dpi=150)
plt.close()


# ── 4. Data-Ink Ratio: Before & After ──────────────────────
# Demonstrate the contrast between a cluttered chart and a
# clean redesign of the same data.

# Count vehicle origins
origin_counts = mpg["origin"].value_counts().sort_values()

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))

# "Before" — heavy gridlines, unnecessary fills, vertical bars
ax1.bar(
    origin_counts.index,
    origin_counts.values,
    color="skyblue",
    edgecolor="black",
    linewidth=1.5,
)
ax1.set_facecolor("#F0F0F0")
ax1.grid(True, which="both", linewidth=0.8, color="grey")
ax1.set_title("Before: Cluttered Design", fontweight="bold")
ax1.set_ylabel("Count")

# "After" — clean horizontal bars with direct labels
ax2.barh(
    origin_counts.index,
    origin_counts.values,
    color="#1565C0",
    height=0.6,
)
for i, (val, name) in enumerate(zip(origin_counts.values, origin_counts.index)):
    ax2.text(val + 3, i, str(val), va="center", fontsize=10)
ax2.set_title("After: Clean Design", fontweight="bold")
ax2.set_xlabel("Count")
ax2.spines["top"].set_visible(False)
ax2.spines["right"].set_visible(False)
ax2.grid(axis="x", alpha=0.3)

fig.suptitle(
    "Data-Ink Ratio: Removing Chartjunk",
    fontsize=13, fontweight="bold",
)
plt.tight_layout()
plt.savefig("output/data_ink_comparison_py.png", dpi=150)
plt.close()

print("── All plots saved to output/ directory ──")

File: latest/week1/workshop1_m1/W01_M01_demo.R
# ============================================================
# Workshop 1 — Module 1: Why Visualize?
# R Demonstration Script
# Data Visualization for Data Scientists — UNYT Tirana
# ============================================================

# ── 0. Environment Setup ────────────────────────────────────
# install.packages(c("tidyverse", "datasauRus", "patchwork"))
library(tidyverse)
library(datasauRus)
library(patchwork)

# ── 1. Anscombe's Quartet ───────────────────────────────────
# Anscombe's quartet ships with base R; four datasets with
# identical summary statistics but completely different shapes.

data(anscombe)

# Compute summary statistics to prove they match
anscombe_summary <- anscombe |>
  pivot_longer(
    everything(),
    names_to  = c(".value", "set"),
    names_pattern = "(.)(.)"
  ) |>
  group_by(set) |>
  summarise(
    mean_x = round(mean(x), 2),
    mean_y = round(mean(y), 2),
    sd_x   = round(sd(x), 2),
    sd_y   = round(sd(y), 2),
    cor_xy = round(cor(x, y), 3),
    .groups = "drop"
  )

cat("── Anscombe's Quartet: Summary Statistics ──\n")
print(anscombe_summary)
# Observe: all four sets yield nearly identical values

# Reshape for faceted plotting
anscombe_long <- anscombe |>
  pivot_longer(
    everything(),
    names_to  = c(".value", "set"),
    names_pattern = "(.)(.)"
  )

# Faceted scatterplot with regression lines
p_anscombe <- ggplot(anscombe_long, aes(x = x, y = y)) +
  geom_point(size = 2.5, color = "#1565C0") +
  geom_smooth(
    method = "lm", se = FALSE,
    color = "#E53935", linewidth = 0.8
  ) +
  facet_wrap(~set, ncol = 2,
             labeller = labeller(set = c(
               "1" = "Dataset I",
               "2" = "Dataset II",
               "3" = "Dataset III",
               "4" = "Dataset IV"
             ))) +
  labs(
    title    = "Anscombe's Quartet",
    subtitle = "Same statistics, different stories",
    x = "x", y = "y"
  ) +
  theme_minimal(base_size = 12)

print(p_anscombe)
ggsave("output/anscombe_quartet.png", p_anscombe,
       width = 8, height = 6, dpi = 150)


# ── 2. Datasaurus Dozen ────────────────────────────────────
# Matejka & Fitzmaurice (2017) extended Anscombe's idea to
# 13 datasets — including a dinosaur — all with identical
# means, standard deviations, and correlations.

# Verify identical statistics
dino_stats <- datasaurus_dozen |>
  group_by(dataset) |>
  summarise(
    mean_x = round(mean(x), 1),
    mean_y = round(mean(y), 1),
    sd_x   = round(sd(x), 1),
    sd_y   = round(sd(y), 1),
    cor_xy = round(cor(x, y), 2),
    .groups = "drop"
  )

cat("\n── Datasaurus Dozen: Summary Statistics ──\n")
print(dino_stats)

# Faceted scatterplot: 13 shapes, one set of statistics
p_datasaurus <- ggplot(datasaurus_dozen,
                       aes(x = x, y = y)) +
  geom_point(alpha = 0.5, size = 0.6, color = "#2E7D32") +
  facet_wrap(~dataset, ncol = 5) +
  labs(
    title    = "The Datasaurus Dozen",
    subtitle = "13 datasets, same summary stats (Matejka & Fitzmaurice, 2017)",
    caption  = "Source: datasauRus package"
  ) +
  theme_minimal(base_size = 8) +
  theme(
    strip.text = element_text(face = "bold", size = 7),
    plot.title = element_text(face = "bold")
  )

print(p_datasaurus)
ggsave("output/datasaurus_dozen.png", p_datasaurus,
       width = 10, height = 6, dpi = 150)


# ── 3. First Plot: mpg Dataset ─────────────────────────────
# A quick preview of ggplot2's layered grammar using the
# built-in mpg dataset (fuel economy from 1999–2008).

p_mpg <- ggplot(mpg, aes(x = displ, y = hwy, color = class)) +
  geom_point(size = 2, alpha = 0.7) +
  labs(
    title = "Engine Displacement vs Highway Fuel Economy",
    x     = "Engine Displacement (litres)",
    y     = "Highway MPG",
    color = "Vehicle Class"
  ) +
  theme_minimal(base_size = 12)

print(p_mpg)
ggsave("output/mpg_scatter.png", p_mpg,
       width = 8, height = 5, dpi = 150)


# ── 4. Data-Ink Ratio: Before & After ──────────────────────
# Demonstrate chartjunk removal using the same data

# "Before" — heavy gridlines, unnecessary decoration
p_before <- ggplot(mpg, aes(x = class)) +
  geom_bar(fill = "skyblue", color = "black", linewidth = 1) +
  labs(title = "Vehicle Counts by Class (Cluttered)") +
  theme(
    panel.grid.major = element_line(color = "grey60", linewidth = 0.8),
    panel.grid.minor = element_line(color = "grey80", linewidth = 0.5),
    panel.background = element_rect(fill = "grey95"),
    axis.text.x = element_text(angle = 45, hjust = 1, size = 10)
  )

# "After" — clean, sorted, horizontal
mpg_counts <- mpg |>
  count(class) |>
  mutate(class = fct_reorder(class, n))

p_after <- ggplot(mpg_counts, aes(x = n, y = class)) +
  geom_col(fill = "#1565C0", width = 0.7) +
  geom_text(aes(label = n), hjust = -0.2, size = 3.5) +
  labs(
    title = "Vehicle Counts by Class (Clean)",
    x = "Count", y = NULL
  ) +
  theme_minimal(base_size = 12) +
  theme(panel.grid.major.y = element_blank())

# Side-by-side comparison using patchwork
p_comparison <- p_before + p_after +
  plot_annotation(
    title = "Data-Ink Ratio: Before & After",
    subtitle = "Removing chartjunk improves communication"
  )

print(p_comparison)
ggsave("output/data_ink_comparison.png", p_comparison,
       width = 12, height = 5, dpi = 150)

cat("\n── All plots saved to output/ directory ──\n")
